pkg/controller/config/dns.go

```diff
@@ -40,17 +42,65 @@ func (h *configHandler) Handle(req router.Request, resp router.Response) error {
 		return err
 	}
 
+	domain, err := h.handleSecret(req, resp, cfg)
+	if err != nil {
+		return err
+	}
+
+	if err = h.handleIngress(req, resp, domain, cfg.AcornDNS, cfg.IngressClassName); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// purgeRecordsIfDisabling checks if we are transitioning AcorDNS from an enabled state to a disabled state and if so calls the
+// acorn DNS service to purge all records for the domain. It is expected that string variable returned by this function
+// will be set as the labels.AcornDNSState annotation on the acorn-dns secret
+func purgeRecordsIfDisabling(req router.Request, domain string, cfg *apiv1.Config, dnsSecret *corev1.Secret,
+	token string, dnsClient dns.Client) (string, error) {
+	var state string
+	if domain == "" {
+		state = *cfg.AcornDNS
+	} else {
+		// The config object includes the acorn domain in the list of ClusterDomains if we are in an enabled state.
+		if slices.Contains(cfg.ClusterDomains, domain) {
+			state = "enabled"
+		} else {
+			state = "disabled"
+		}
+	}
+
+	// purge the records if the current state is disabled, but that isn't the state recorded on the dnsSecret
+	// If this is the first time the handler is being called, then the domain and token will be blank and thus we
+	// shouldn't call purge
+	if strings.EqualFold(state, "disabled") && dnsSecret.Annotations[labels.AcornDNSState] != "disabled" {
+		if domain != "" && token != "" {
+			if err := dnsClient.PurgeRecords(*cfg.AcornDNSEndpoint, domain, token); err != nil {
+				if dns.IsDomainAuthError(err) {
+					if err := dns.ClearDNSToken(req.Ctx, req.Client, dnsSecret); err != nil {
+						return "", err
+					}
+				}
+				return "", err
+			}
+		}
+	}
+	return state, nil
+}
+
+func (h *configHandler) handleSecret(req router.Request, resp router.Response, cfg *apiv1.Config) (string, error) {
```

Comment:
Broke this out into its own function to reduce the complexity burden on the `Handle` function.